File: scripts/detection/check_upstream_recovery.mjs
#!/usr/bin/env node
// scripts/check_upstream_recovery.mjs
/**
 * Upstream recovery detection — standalone aggregator (Phase A).
 *
 * 既存 signals を読み取り、`upstream-recovery-status.json` を派生する:
 *   - `EN_SOURCE_PATCHES` + `preprocessEnHtml(patchCoverage)` → per-patch Axis A status
 *   - `source-sync-status.json.pages[].fetchStatus` → per-exclusion Axis A status
 *   - 各 entry の `reviewAfter` との cadence 比較 → per-entry Axis B status
 *
 * Non-blocking: process.exit(0) unconditionally. consumers (sticky PR comment /
 * detection_reports / sourceSyncHealth managed issue) が JSON を読んで判断する。
 *
 * Architecture invariants:
 *   - `check_source_parity.mjs` / `scripts/lib/parity_*.mjs` には触らない
 *     (parity gate の挙動を変更しない — baseline=0 regression 禁止)
 *   - 新 detector / issue type / workflow は追加しない (既存 infra の拡張のみ)
 *
 * Plan: docs/SYSTEM_SPEC.md §システム不変量
 * Spec: docs/SYSTEM_SPEC.md §システム不変量
 *
 * @module check_upstream_recovery
 */

import { readFileSync, writeFileSync, existsSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { ROOT_DIR } from '../lib/project.mjs';
import {
  EN_SOURCE_PATCHES,
  createEnSourcePatchCoverage,
} from '../lib/en_source_patches.mjs';
import { SOURCE_SYNC_EXCLUSIONS } from '../lib/source_sync_exclusions.mjs';
import { preprocessEnHtml } from '../lib/turndown.mjs';

const SNAPSHOTS_ROOT = path.join(ROOT_DIR, 'snapshots', 'en', 'content');
const SOURCE_SYNC_STATUS_PATH = path.join(ROOT_DIR, 'source-sync-status.json');
const OUTPUT_PATH = path.join(ROOT_DIR, 'upstream-recovery-status.json');

const MS_PER_DAY = 24 * 60 * 60 * 1000;

/**
 * Days between `dateStr` (YYYY-MM-DD) and `now`.
 *
 * - Positive when `dateStr` is in the past (overdue).
 * - `0` on the same UTC day.
 * - Negative when `dateStr` is in the future (not yet due).
 * - `0` for invalid / missing dates (fail-safe — callers use `> 0` to mean overdue).
 *
 * `new Date('YYYY-MM-DD')` is parsed as UTC midnight per ES2015+, so the result
 * is timezone-independent.
 *
 * @param {string | null | undefined} dateStr
 * @param {number} nowMs
 */
export function daysSince(dateStr, nowMs = Date.now()) {
  if (typeof dateStr !== 'string' || dateStr.length === 0) return 0;
  const parsed = new Date(dateStr).getTime();
  if (!Number.isFinite(parsed)) return 0;
  return Math.floor((nowMs - parsed) / MS_PER_DAY);
}

/**
 * Days between `now` and `dateStr`. Positive when dateStr is in the future.
 * Mirror of `daysSince` for the opposite sign — used to show "days until review".
 *
 * @param {string | null | undefined} dateStr
 * @param {number} nowMs
 */
export function daysUntil(dateStr, nowMs = Date.now()) {
  if (typeof dateStr !== 'string' || dateStr.length === 0) return null;
  const parsed = new Date(dateStr).getTime();
  if (!Number.isFinite(parsed)) return null;
  return Math.floor((parsed - nowMs) / MS_PER_DAY);
}

/**
 * Overdue predicate with same-day semantics aligned to
 * `scripts/check_patch_review_cadence.mjs::evaluatePatchReview` (Codex C1):
 *
 *   - `reviewAfter` is a YYYY-MM-DD string, parsed as UTC midnight by
 *     `new Date(...)` per ES2015+.
 *   - Overdue when `nowMs > reviewAfterMs` **strictly** — the review day
 *     itself (nowMs === reviewAfterMs) is *not* overdue (inclusive boundary).
 *   - Invalid / missing dates fail-safe to `false`.
 *
 * Using this helper instead of `daysSince(...) > 0` keeps both detectors
 * aligned even when nowMs lands partway through the review day (e.g. 12:00
 * UTC on reviewAfter — cadence would flag overdue, the old daysSince check
 * would not).
 *
 * @param {string | null | undefined} dateStr
 * @param {number} nowMs
 * @returns {boolean}
 */
export function isReviewOverdue(dateStr, nowMs = Date.now()) {
  if (typeof dateStr !== 'string' || dateStr.length === 0) return false;
  const parsed = new Date(dateStr).getTime();
  if (!Number.isFinite(parsed)) return false;
  return nowMs > parsed;
}

/**
 * Load existing source-sync-status.json (if present). Absent file is a legitimate
 * local-development state and must degrade gracefully — callers treat missing
 * data as `fetchStatus: 'unknown'`.
 *
 * @returns {{ pages: Array<{ slug: string, fetchStatus?: string }> } | null}
 */
function loadSourceSyncStatus(filePath = SOURCE_SYNC_STATUS_PATH) {
  if (!existsSync(filePath)) return null;
  try {
    return JSON.parse(readFileSync(filePath, 'utf8'));
  } catch (err) {
    console.warn(
      `[upstream-recovery] failed to parse ${path.basename(filePath)}: ${err.message}`,
    );
    return null;
  }
}

/**
 * Collect unique slugs across all patches (slug-driven loop per plan Task 2).
 * `preprocessEnHtml` applies every applicable patch for the given slug in a
 * single pass, so iterating slugs (not patches) avoids duplicate work.
 */
function uniquePatchSlugs(patches = EN_SOURCE_PATCHES) {
  const set = new Set();
  for (const patch of patches) {
    for (const slug of patch.slugs) set.add(slug);
  }
  return set;
}

/**
 * Axis A — EN upstream 修正検知 (en_patches).
 *
 * Returns per-patch rows with `statusA` (active/stale/unknown), `statusB`
 * (current/overdue), hits count, and cadence metadata.
 *
 * **Note on the `patches` parameter (Codex C2):** this argument is used only
 * for the outer enumeration loop (slug collection + result shaping). The
 * actual hit detection happens inside `preprocessEnHtml`, which reads the
 * *live* `EN_SOURCE_PATCHES` constant — not the injected `patches`. Unit
 * tests therefore inject synthetic patches to exercise stale / unknown
 * branches only; the `active` branch requires real registered slugs and is
 * covered by `en_source_patches_integration.test.mjs`. Production callers
 * should omit `patches` to use the default (live registry) and stay
 * consistent with the hit-detection layer.
 *
 * @param {number} nowMs
 * @param {string} snapshotsRoot
 * @param {ReadonlyArray<object>} [patches] — override the enumeration list
 *   (tests only; does NOT change hit detection — see note above)
 */
export function computeEnPatchStatus({
  nowMs = Date.now(),
  snapshotsRoot = SNAPSHOTS_ROOT,
  patches = EN_SOURCE_PATCHES,
} = {}) {
  const coverage = createEnSourcePatchCoverage();
  const slugs = uniquePatchSlugs(patches);
  const snapshotSeen = new Set();
  for (const slug of slugs) {
    const snapshotPath = path.join(snapshotsRoot, `${slug}.html`);
    if (!existsSync(snapshotPath)) continue;
    const raw = readFileSync(snapshotPath, 'utf8');
    try {
      preprocessEnHtml(raw, { slug, patchCoverage: coverage });
      snapshotSeen.add(slug);
    } catch (err) {
      console.warn(
        `[upstream-recovery] preprocessEnHtml failed for slug=${slug}: ${err.message}`,
      );
    }
  }
  const snap = coverage.snapshot();
  return patches.map((patch) => {
    const status = snap.byPatchIdStatus[patch.id] ?? { matched: false, hits: 0 };
    // A patch is considered `unknown` when NONE of its registered slugs had a
    // readable snapshot in this run (can't tell stale vs active). Otherwise
    // matched=true → active, matched=false → stale (upstream likely fixed).
    const anySnapshotSeen = patch.slugs.some((s) => snapshotSeen.has(s));
    let statusA;
    if (!anySnapshotSeen) {
      statusA = 'unknown';
    } else {
      statusA = status.matched ? 'active' : 'stale';
    }
    // Use isReviewOverdue (not daysSince(...) > 0) so same-day semantics
    // match check_patch_review_cadence.mjs (Codex C1).
    const statusB = isReviewOverdue(patch.reviewAfter, nowMs) ? 'overdue' : 'current';
    return {
      id: patch.id,
      mechanism: 'en_source_patches',
      slugs: [...patch.slugs],
      statusA,
      statusB,
      hits: status.hits,
      addedAt: patch.addedAt ?? null,
      reviewAfter: patch.reviewAfter ?? null,
      daysUntilReview: daysUntil(patch.reviewAfter, nowMs),
    };
  });
}

/**
 * Axis A — EN upstream 修正検知 (sync_exclusions).
 *
 * Reads existing `fetchStatus` signal from `source-sync-status.json`. Missing
 * file / missing entry degrades to `unknown` (local dev / PR CI without the
 * artifact). `excluded-recovered` → stale, `excluded-broken` → active.
 *
 * @param {number} nowMs
 */
export function computeSyncExclusionStatus({
  nowMs = Date.now(),
  exclusions = SOURCE_SYNC_EXCLUSIONS,
  // NOTE: the I/O-bearing default `loadSourceSyncStatus()` is evaluated once
  // per call-without-argument. All tests inject `sourceSyncStatus` directly
  // (no fs touches from unit tests); the default is reached only from the
  // CLI path (buildUpstreamRecoveryStatus → computeSyncExclusionStatus).
  sourceSyncStatus = loadSourceSyncStatus(),
} = {}) {
  const pages = Array.isArray(sourceSyncStatus?.pages) ? sourceSyncStatus.pages : [];
  const pageBySlug = new Map();
  for (const page of pages) {
    if (typeof page?.slug === 'string') pageBySlug.set(page.slug, page);
  }
  return Object.entries(exclusions).map(([slug, entry]) => {
    const page = pageBySlug.get(slug);
    const fetchStatus = page?.fetchStatus ?? 'unknown';
    let statusA;
    if (fetchStatus === 'excluded-recovered') {
      statusA = 'stale';
    } else if (fetchStatus === 'excluded-broken') {
      statusA = 'active';
    } else {
      statusA = 'unknown';
    }
    // isReviewOverdue matches check_patch_review_cadence.mjs same-day semantics
    // (Codex C1). Missing reviewAfter is treated as not-overdue (fail-safe).
    const statusB = isReviewOverdue(entry.reviewAfter, nowMs) ? 'overdue' : 'current';
    return {
      slug,
      mechanism: 'source_sync_exclusions',
      statusA,
      statusB,
      fetchStatus,
      addedAt: entry.addedAt ?? null,
      reviewAfter: entry.reviewAfter ?? null,
      daysUntilReview: entry.reviewAfter ? daysUntil(entry.reviewAfter, nowMs) : null,
    };
  });
}

/**
 * Build the full `upstream-recovery-status.json` payload in memory. Caller is
 * responsible for persistence. Exported for unit tests and downstream tools
 * (detection_reports, sticky PR comment in Phase B).
 *
 * @param {object} [options]
 */
export function buildUpstreamRecoveryStatus(options = {}) {
  const nowMs = options.nowMs ?? Date.now();
  const enPatches = computeEnPatchStatus({ ...options, nowMs });
  const syncExclusions = computeSyncExclusionStatus({ ...options, nowMs });
  const allEntries = [...enPatches, ...syncExclusions];
  const staleCount = allEntries.filter((e) => e.statusA === 'stale').length;
  const overdueCount = allEntries.filter((e) => e.statusB === 'overdue').length;
  const unknownCount = allEntries.filter((e) => e.statusA === 'unknown').length;
  const totalEntries = allEntries.length;
  const activeEntries = totalEntries - staleCount - unknownCount;
  return {
    schemaVersion: 1,
    generatedAt: new Date(nowMs).toISOString(),
    summary: {
      totalEntries,
      activeEntries,
      staleEntries: staleCount, // Axis A signal
      overdueEntries: overdueCount, // Axis B signal
      unknownEntries: unknownCount,
    },
    mechanisms: {
      en_source_patches: enPatches,
      source_sync_exclusions: syncExclusions,
    },
  };
}

export function runCheckUpstreamRecovery({
  outputPath = OUTPUT_PATH,
  stdout = console.log,
  ...options
} = {}) {
  const payload = buildUpstreamRecoveryStatus(options);
  writeFileSync(outputPath, JSON.stringify(payload, null, 2) + '\n');
  stdout(
    `[upstream-recovery] total=${payload.summary.totalEntries} ` +
      `active=${payload.summary.activeEntries} ` +
      `stale=${payload.summary.staleEntries} ` +
      `overdue=${payload.summary.overdueEntries} ` +
      `unknown=${payload.summary.unknownEntries} ` +
      `→ ${path.relative(ROOT_DIR, outputPath)}`,
  );
  return payload;
}

const isDirectRun =
  process.argv[1] && fileURLToPath(import.meta.url) === path.resolve(process.argv[1]);
if (isDirectRun) {
  try {
    runCheckUpstreamRecovery();
    process.exit(0); // Non-blocking: consumers decide via JSON.
  } catch (err) {
    console.error(`[upstream-recovery] unexpected failure: ${err.message}`);
    process.exit(0); // Still non-blocking — don't break unrelated CI.
  }
}

File: scripts/python/src/testim_parity/detection/check_upstream_recovery.py
"""``scripts/detection/check_upstream_recovery.mjs`` の Python port。

既存 signal を aggregate して ``upstream-recovery-status.json`` を派生する:

- ``EN_SOURCE_PATCHES`` + ``preprocess_en_html(patch_coverage)`` → per-patch Axis A
- ``source-sync-status.json.pages[].fetchStatus`` → per-exclusion Axis A
- 各 entry の ``reviewAfter`` との cadence 比較 → per-entry Axis B

Non-blocking (exit 0 常時)。consumers (sticky PR comment / detection_reports /
sourceSyncHealth managed issue) が JSON を読んで判断する。
"""

from __future__ import annotations

import argparse
import json
import sys
from collections.abc import Mapping, Sequence
from datetime import UTC, datetime
from pathlib import Path
from typing import IO, Any

from ..en_source_patches import EN_SOURCE_PATCHES, create_en_source_patch_coverage
from ..preprocess_en import preprocess_en_html
from ..project import ROOT_DIR
from ..sync_exclusions import SOURCE_SYNC_EXCLUSIONS

__all__ = [
    "build_upstream_recovery_status",
    "compute_en_patch_status",
    "compute_sync_exclusion_status",
    "days_since",
    "days_until",
    "is_review_overdue",
    "main",
    "run_check_upstream_recovery",
]


_SNAPSHOTS_ROOT: Path = ROOT_DIR / "snapshots" / "en" / "content"
_SOURCE_SYNC_STATUS_PATH: Path = ROOT_DIR / "source-sync-status.json"
_OUTPUT_PATH: Path = ROOT_DIR / "upstream-recovery-status.json"

_MS_PER_DAY: int = 24 * 60 * 60 * 1000


def _now_ms() -> int:
    return int(datetime.now(tz=UTC).timestamp() * 1000)


def _iso_to_ms(date_str: Any) -> float | None:
    if not isinstance(date_str, str) or not date_str:
        return None
    try:
        dt = datetime.fromisoformat(date_str).replace(tzinfo=UTC)
    except ValueError:
        return None
    return dt.timestamp() * 1000


def days_since(date_str: Any, now_ms: int | None = None) -> int:
    """``date_str`` から now までの日数 (``> 0`` で overdue)。"""
    if now_ms is None:
        now_ms = _now_ms()
    parsed = _iso_to_ms(date_str)
    if parsed is None:
        return 0
    return int((now_ms - parsed) // _MS_PER_DAY)


def days_until(date_str: Any, now_ms: int | None = None) -> int | None:
    """now から ``date_str`` までの日数。invalid / 欠損で ``None``。"""
    if now_ms is None:
        now_ms = _now_ms()
    parsed = _iso_to_ms(date_str)
    if parsed is None:
        return None
    return int((parsed - now_ms) // _MS_PER_DAY)


def is_review_overdue(date_str: Any, now_ms: int | None = None) -> bool:
    """``check_patch_review_cadence`` と同じ strictly-greater セマンティクス。"""
    if now_ms is None:
        now_ms = _now_ms()
    parsed = _iso_to_ms(date_str)
    if parsed is None:
        return False
    return now_ms > parsed


def _load_source_sync_status(file_path: Path = _SOURCE_SYNC_STATUS_PATH) -> Any:
    """``source-sync-status.json`` を読む。不在 / 破損で None (mjs 等価)。"""
    if not file_path.exists():
        return None
    try:
        return json.loads(file_path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as err:
        print(
            f"[upstream-recovery] failed to parse {file_path.name}: {err}",
            file=sys.stderr,
        )
        return None


def _unique_patch_slugs(patches: Sequence[Mapping[str, Any]]) -> set[str]:
    """全 patch の slug union (mjs 等価)。"""
    out: set[str] = set()
    for patch in patches:
        slugs = patch.get("slugs") or []
        for slug in slugs:
            out.add(slug)
    return out


def compute_en_patch_status(
    *,
    now_ms: int | None = None,
    snapshots_root: Path | str = _SNAPSHOTS_ROOT,
    patches: Sequence[Mapping[str, Any]] | None = None,
) -> list[dict[str, Any]]:
    """Axis A: en_source_patches (mjs 等価)。"""
    if now_ms is None:
        now_ms = _now_ms()
    patches_list: Sequence[Mapping[str, Any]] = (
        patches if patches is not None else EN_SOURCE_PATCHES
    )
    coverage = create_en_source_patch_coverage()
    slugs = _unique_patch_slugs(patches_list)
    snapshot_seen: set[str] = set()
    root = Path(snapshots_root)
    for slug in slugs:
        snapshot_path = root / f"{slug}.html"
        if not snapshot_path.exists():
            continue
        # ``en_source_patches`` find strings intentionally preserve MadCap CRLF
        # boundaries. ``Path.read_text`` normalizes newlines on some platforms and
        # can make an active patch look stale, so mirror check_source_parity.
        raw = snapshot_path.read_bytes().decode("utf-8")
        try:
            preprocess_en_html(raw, slug=slug, patch_coverage=coverage)
            snapshot_seen.add(slug)
        except Exception as err:
            print(
                f"[upstream-recovery] preprocess_en_html failed for slug={slug}: {err}",
                file=sys.stderr,
            )

    snap = coverage["snapshot"]()
    by_patch_id_status = snap.get("byPatchIdStatus", {})

    result: list[dict[str, Any]] = []
    for patch in patches_list:
        status = by_patch_id_status.get(patch.get("id")) or {"matched": False, "hits": 0}
        patch_slugs = list(patch.get("slugs") or [])
        any_snapshot_seen = any(s in snapshot_seen for s in patch_slugs)
        if not any_snapshot_seen:
            status_a = "unknown"
        else:
            status_a = "active" if status.get("matched") else "stale"
        status_b = "overdue" if is_review_overdue(patch.get("reviewAfter"), now_ms) else "current"
        result.append(
            {
                "id": patch.get("id"),
                "mechanism": "en_source_patches",
                "slugs": patch_slugs,
                "statusA": status_a,
                "statusB": status_b,
                "hits": status.get("hits", 0),
                "addedAt": patch.get("addedAt"),
                "reviewAfter": patch.get("reviewAfter"),
                "daysUntilReview": days_until(patch.get("reviewAfter"), now_ms),
            }
        )
    return result


def compute_sync_exclusion_status(
    *,
    now_ms: int | None = None,
    exclusions: Mapping[str, Mapping[str, Any]] | None = None,
    source_sync_status: Any = None,
) -> list[dict[str, Any]]:
    """Axis A: source_sync_exclusions (mjs 等価)。"""
    if now_ms is None:
        now_ms = _now_ms()
    if exclusions is None:
        exclusions = SOURCE_SYNC_EXCLUSIONS
    if source_sync_status is None:
        source_sync_status = _load_source_sync_status()
    pages = (source_sync_status or {}).get("pages") or []
    if not isinstance(pages, list):
        pages = []
    page_by_slug: dict[str, Mapping[str, Any]] = {}
    for page in pages:
        slug = (page or {}).get("slug")
        if isinstance(slug, str):
            page_by_slug[slug] = page

    result: list[dict[str, Any]] = []
    for slug, entry in exclusions.items():
        page = page_by_slug.get(slug)
        fetch_status = (page or {}).get("fetchStatus") or "unknown"
        if fetch_status == "excluded-recovered":
            status_a = "stale"
        elif fetch_status == "excluded-broken":
            status_a = "active"
        else:
            status_a = "unknown"
        status_b = "overdue" if is_review_overdue(entry.get("reviewAfter"), now_ms) else "current"
        result.append(
            {
                "slug": slug,
                "mechanism": "source_sync_exclusions",
                "statusA": status_a,
                "statusB": status_b,
                "fetchStatus": fetch_status,
                "addedAt": entry.get("addedAt"),
                "reviewAfter": entry.get("reviewAfter"),
                "daysUntilReview": (
                    days_until(entry.get("reviewAfter"), now_ms)
                    if entry.get("reviewAfter")
                    else None
                ),
            }
        )
    return result


def build_upstream_recovery_status(
    *,
    now_ms: int | None = None,
    snapshots_root: Path | str = _SNAPSHOTS_ROOT,
    patches: Sequence[Mapping[str, Any]] | None = None,
    exclusions: Mapping[str, Mapping[str, Any]] | None = None,
    source_sync_status: Any = None,
) -> dict[str, Any]:
    """``upstream-recovery-status.json`` payload を組み立てる (mjs 等価)。"""
    if now_ms is None:
        now_ms = _now_ms()

    en_patches = compute_en_patch_status(
        now_ms=now_ms, snapshots_root=snapshots_root, patches=patches
    )
    sync_exclusions = compute_sync_exclusion_status(
        now_ms=now_ms, exclusions=exclusions, source_sync_status=source_sync_status
    )
    all_entries = en_patches + sync_exclusions
    stale = sum(1 for e in all_entries if e.get("statusA") == "stale")
    overdue = sum(1 for e in all_entries if e.get("statusB") == "overdue")
    unknown = sum(1 for e in all_entries if e.get("statusA") == "unknown")
    total = len(all_entries)
    active = total - stale - unknown

    # mjs ``new Date(nowMs).toISOString()`` 等価。
    dt = datetime.fromtimestamp(now_ms / 1000, tz=UTC)
    generated_at = dt.strftime("%Y-%m-%dT%H:%M:%S.") + f"{dt.microsecond // 1000:03d}Z"

    return {
        "schemaVersion": 1,
        "generatedAt": generated_at,
        "summary": {
            "totalEntries": total,
            "activeEntries": active,
            "staleEntries": stale,
            "overdueEntries": overdue,
            "unknownEntries": unknown,
        },
        "mechanisms": {
            "en_source_patches": en_patches,
            "source_sync_exclusions": sync_exclusions,
        },
    }


def run_check_upstream_recovery(
    *,
    output_path: Path | str = _OUTPUT_PATH,
    stdout: IO[str] | None = None,
    now_ms: int | None = None,
    snapshots_root: Path | str = _SNAPSHOTS_ROOT,
    patches: Sequence[Mapping[str, Any]] | None = None,
    exclusions: Mapping[str, Mapping[str, Any]] | None = None,
    source_sync_status: Any = None,
) -> dict[str, Any]:
    """payload を書き出し summary line を stdout に出す (mjs 等価)。"""
    out_stream = stdout if stdout is not None else sys.stdout
    output = Path(output_path)

    payload = build_upstream_recovery_status(
        now_ms=now_ms,
        snapshots_root=snapshots_root,
        patches=patches,
        exclusions=exclusions,
        source_sync_status=source_sync_status,
    )
    output.write_text(
        json.dumps(payload, ensure_ascii=False, indent=2) + "\n",
        encoding="utf-8",
    )
    summary = payload["summary"]
    try:
        rel = output.relative_to(ROOT_DIR)
    except ValueError:
        rel = output
    print(
        f"[upstream-recovery] total={summary['totalEntries']} "
        f"active={summary['activeEntries']} "
        f"stale={summary['staleEntries']} "
        f"overdue={summary['overdueEntries']} "
        f"unknown={summary['unknownEntries']} "
        f"→ {rel}",
        file=out_stream,
    )
    return payload


def main(argv: list[str] | None = None) -> int:
    """CLI エントリポイント (常に exit 0 = non-blocking)。"""
    parser = argparse.ArgumentParser(
        description="Derive upstream-recovery-status.json from existing signals"
    )
    _ = parser.parse_args(argv)
    try:
        run_check_upstream_recovery()
    except Exception as err:  # pragma: no cover - defensive
        print(f"[upstream-recovery] unexpected failure: {err}", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())
